feat: add `rhsm_enabled` for Red Hat Enterprise Linux

nesies opened the pull request:
<!--

In order to have a good experience with our community, we recommend that you read the [contributing guidelines](https://github.com/vmware-samples/packer-examples-for-vsphere/blob/main/CONTRIBUTING.md) for making a pull request.

-->

# <!-- Intentionally left blank. -->

## Summary of Pull Request

Add a new bool variable `rhsm_enabled`, to enable or not Red Hat Subscription Manager

Default: `true` to maintain backward compatibility.

<!--
    Please provide a clear and concise description of the pull request.
-->

## Type of Pull Request

<!--
    Please check the one that applies to this pull request using "[x]".
-->

- [ ] This is a bugfix. `type/bug`
- [X] This is an enhancement or feature. `type/feature` or `type/enhancement`
- [ ] This is a documentation update. `type/docs`
- [ ] This is a refactoring update. `type/refactor`
- [ ] This is a chore. `type/chore`
- [ ] This is something else.
      Please describe:

## Related to Existing Issues

Closes #977

## Test and Documentation Coverage

<!--
    Please check the one that applies to this pull request using "[x]".
-->

- [ ] Tests have been completed.
- [X] Documentation has been added or updated.

## Breaking Changes?

<!--
    Please check the one that applies to this pull request using "[x]".
-->

- [ ] Yes, there are breaking changes.
- [X] No, there are no breaking changes.

<!--
    If this pull request contains a breaking change, please describe the impact and mitigation path.
-->

tenthirtyam commented on builds/linux/rhel/8/linux-rhel.pkr.hcl:
```diff
@@ -231,6 +232,7 @@ build {
       "--extra-vars", "ansible_username=${var.ansible_username}",
       "--extra-vars", "ansible_key='${var.ansible_key}'",
       "--extra-vars", "enable_cloudinit=${var.vm_guest_os_cloudinit}",
+      "--extra-vars", "{\"rhsm_enabled\": ${var.rhsm_enabled}}",
```
Formatting is different than a similar above. Any reason?

nesies replied:
_rhsm_enabled_ is a boolean, if we use _-e rhsm_enable=_ it will be interpreted as a _string_ by _ansible_

tenthirtyam replied:
Ah, good point! Might need to modify the enable_cloudinit in a follow-up PR to do the same.

nesies replied:
in ansible, the conditionnals are done againt "true" or "false" strings values, so there is no fonctionnal bug.

But it is "not clean"

tenthirtyam commented on builds/linux/rhel/8/variables.pkr.hcl:
```diff
@@ -12,6 +12,12 @@
 
 // Red Hat Subscription Manager Credentials
 
+variable "rhsm_enabled" {
+  type        = bool
+  description = "Enable Red Hat Subscription Manager."
+  sensitive   = false
+}
```
```suggestion
variable "rhsm_enabled" {
  type        = bool
  description = "Enable Red Hat Subscription Manager."
}
```

nesies replied:
sensitive removed

tenthirtyam commented on builds/linux/rhel/9/variables.pkr.hcl:
```diff
@@ -10,8 +10,24 @@
 //  BLOCK: variable
 //  Defines the input variables.
 
+// Additional yum repositories
```
```suggestion
// Additional Repositories
```

nesies replied:
yum_repository should have been in the pull request

tenthirtyam commented on builds/linux/rhel/9/variables.pkr.hcl:
```diff
@@ -10,8 +10,24 @@
+variable "rhsm_enabled" {
+  type        = bool
+  description = "Enable Red Hat Subscription Manager."
+  sensitive   = false
```
```suggestion
```

tenthirtyam commented on builds/rhsm.pkrvars.hcl.example:
```diff
@@ -8,5 +8,6 @@
 */
 
 // Red Hat Subscription Manager Credentials
+rhsm_enabled   = true
```
```suggestion

rhsm_enabled   = true
```

tenthirtyam commented on builds/linux/rhel/9/variables.pkr.hcl:
```diff
@@ -10,8 +10,24 @@
+    name = string
+    url = string
+  }))
+  description = "Additional yum repositories"
```
```suggestion
  description = "Additional repositories."
```

nesies replied:
yum_repositories should not have been in this pull, it removed it

tenthirtyam commented on builds/linux/rhel/9/variables.pkr.hcl:
```diff
@@ -10,8 +10,24 @@
 
+// Additional yum repositories
+variable "yum_repositories" {
+  type        = list(object({
```
```suggestion
  type = list(object({
```

tenthirtyam commented on builds/linux/rhel/9/variables.pkr.hcl:
```diff
@@ -10,8 +10,24 @@
 //  Defines the input variables.
 
+// Additional yum repositories
+variable "yum_repositories" {
```
I don't see where this is being used for this enhancement.

nesies replied:
my bad, removed

nesies committed:
feat: add `rhsm_enabled` to enable or disable Red Hat

Add a new bool variable `rhsm_enabled`, to enable or not Red Hat Subscription Manager

Default: `true` to maintain backward compatibility.

Closes #977

Signed-off-by: Fabien SEISEN <seisen@gmail.com>

tenthirtyam commented:
Hi @nesies - please amend your commit with [the required sign-off](https://github.com/vmware-samples/packer-examples-for-vsphere/pull/980/checks?check_run_id=32644168768). Once updated, we'll be able to review and proceed.

Thanks!

**Ryan Johnson**
Distinguished Engineer, VMware by Broadcom

tenthirtyam reviewed (dismissed):
LGTM - added two commits to include in the v0.22.0 release.

tenthirtyam commented on builds/linux/rhel/9/variables.pkr.hcl:
```diff
@@ -459,3 +464,21 @@ variable "additional_packages" {
+// Additional rpm gpg keys
+variable "rpm_gpg_keys" {
+  type        = list(string)
+  description = "Additional rpm gpg keys"
```
```suggestion
  description = "Additional RPM GPG keys."
```

tenthirtyam commented on builds/linux/rhel/9/variables.pkr.hcl:
```diff
@@ -459,3 +464,21 @@ variable "additional_packages" {
   default     = []
 }
+
+// Additional rpm gpg keys
```
```suggestion
// Additional RPM GPG keys.
```

nesies replied:
modified

tenthirtyam commented on builds/linux/rhel/9/linux-rhel.pkrvars.hcl.example:
```diff
@@ -21,3 +21,20 @@ vm_firmware              = "efi-secure"
+  "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release"
+]
+
+yum_repositories = [
+  {
+    "name": "baseos",
+    "url": "http://url/path/to/repo",
+    "install": true
+  },
+  {
+    "name": "appstream",
+    "url": "http://url/path/to/repo",
+    "install": true
+  }
+]
```
I think that this should be defined at a variable and only used if additional repos are enabled enabled.

nesies replied:
commented :)

tenthirtyam commented on builds/linux/rhel/9/linux-rhel.pkrvars.hcl.example:
```diff
@@ -21,3 +21,20 @@ vm_firmware              = "efi-secure"
 iso_content_library_item = "rhel-9.4-x86_64-dvd"
 iso_file                 = "rhel-9.4-x86_64-dvd.iso"
+
+rpm_gpg_keys = [
```
Should this only be used if you want to use specific GPG keys or does this need to be new requirement?

nesies replied:
by default, it seems there is no gpg key in RedHat

tenthirtyam commented on docs/getting-started/configure.md:
```diff
@@ -256,6 +256,8 @@ additional_packages = ["git", "make", "vim"]
 Edit the `config/redhat.pkrvars.hcl` file to configure the credentials for your Red Hat Subscription
 Manager account.
 
+You can also disable Red Hat Subscription Manger by setting `rhsm_enabled = false`.
```
There were additions for other options that are not mentioned here.

nesies replied:
added

nesies replied:
added

tenthirtyam commented on builds/linux/rhel/9/linux-rhel.pkr.hcl:
```diff
@@ -74,6 +75,8 @@ locals {
         lvm        = var.vm_disk_lvm
       })
       additional_packages = join(" ", var.additional_packages)
+      rpm_gpg_keys = var.rpm_gpg_keys
+      yum_repositories = var.yum_repositories
```
Should this be only for RHEL or expanded to the other RHEL-based distributions?

nesies replied:
even if not using RHSM, it can be usefull

tenthirtyam replied:
So this should ideally be added to support the other RHEL-based distributions in the project.

tenthirtyam commented:
Marked as draft until changes are addressed.

tenthirtyam commented:
@nesies this pull request will need some updates for full coverage in other rhel-based machines if including `rpm_gpg_keys` and `yum_repositories`.

nesies commented:
i need some time to see and implement that for other RedHat bases distros, thx
